{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A06_0.sorted.markdup.bam</td>\n",
       "      <td>A06_0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>B06_0.sorted.markdup.bam</td>\n",
       "      <td>B06_0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D06_0.sorted.markdup.bam</td>\n",
       "      <td>D06_0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>F11_1.sorted.markdup.bam</td>\n",
       "      <td>F11_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>D01_1.sorted.markdup.bam</td>\n",
       "      <td>D01_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>F10_1.sorted.markdup.bam</td>\n",
       "      <td>F10_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        file sample  cell_count\n",
       "4   A06_0.sorted.markdup.bam  A06_0           0\n",
       "6   B06_0.sorted.markdup.bam  B06_0           0\n",
       "0   D06_0.sorted.markdup.bam  D06_0           0\n",
       "59  F11_1.sorted.markdup.bam  F11_1           1\n",
       "56  D01_1.sorted.markdup.bam  D01_1           1\n",
       "11  F10_1.sorted.markdup.bam  F10_1           1"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "\n",
    "# generate an empty dataframe to be filled with each sample sequenced in the experiment\n",
    "ntc_sample_df = pd.DataFrame()\n",
    "sc_sample_df = pd.DataFrame()\n",
    "\n",
    "# assign the file path to the experiments sequencing data\n",
    "directory = \"/Volumes/projects_secondary/shen/projects/2022_07_07_RMG2_Data/SHEH_20220214_WGBS_scWGBS/processed_data/scWGBS/analysis/align/\"\n",
    "\n",
    "# iterate over each file in the directory, remove duplicate names, and populate lists with a minimal sample name as well as the cell counts\n",
    "ntc_file_list = []\n",
    "ntc_sample_list = []\n",
    "ntc_cell_count_list = []\n",
    "\n",
    "sc_file_list = []\n",
    "sc_sample_list = []\n",
    "sc_cell_count_list = []\n",
    "\n",
    "for filename in os.listdir(directory):\n",
    "    if \"sorted.markdup.bam.bai\" in filename:\n",
    "        if \"_0\" in filename:\n",
    "            ntc_cell_count_list.append(0)\n",
    "            ntc_file_list.append(filename[:24])\n",
    "            ntc_sample_list.append(filename[:5])\n",
    "        if \"_1\" in filename:\n",
    "            sc_cell_count_list.append(1)\n",
    "            sc_file_list.append(filename[:24])\n",
    "            sc_sample_list.append(filename[:5])       \n",
    "\n",
    "# fill the sample data frame in with the list generated from the experiment directory\n",
    "ntc_sample_df[\"file\"] = ntc_file_list\n",
    "ntc_sample_df[\"sample\"] = ntc_sample_list\n",
    "ntc_sample_df[\"cell_count\"] = ntc_cell_count_list\n",
    "\n",
    "sc_sample_df[\"file\"] = sc_file_list\n",
    "sc_sample_df[\"sample\"] = sc_sample_list\n",
    "sc_sample_df[\"cell_count\"] = sc_cell_count_list\n",
    "\n",
    "# randomly take a subset of the samples sent to sequencing\n",
    "ntc_test_df = ntc_sample_df.sample(frac=0.4)\n",
    "sc_test_df = sc_sample_df.sample(frac=0.05)\n",
    "test_df = pd.concat([ntc_test_df, sc_test_df])\n",
    "\n",
    "# read the data into permanent storage and optionally check the samples in the test dataframe\n",
    "# test_df.to_csv(\"/Users/David.Sokol/Documents/experiments/20220718_scwgbs_contamination_analysis/data/test_df.csv\")\n",
    "test_df #;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_count</th>\n",
       "      <th>chr1</th>\n",
       "      <th>chr2</th>\n",
       "      <th>chr3</th>\n",
       "      <th>chr4</th>\n",
       "      <th>chr5</th>\n",
       "      <th>chr6</th>\n",
       "      <th>chr7</th>\n",
       "      <th>...</th>\n",
       "      <th>chr15</th>\n",
       "      <th>chr16</th>\n",
       "      <th>chr17</th>\n",
       "      <th>chr18</th>\n",
       "      <th>chr19</th>\n",
       "      <th>chr20</th>\n",
       "      <th>chr21</th>\n",
       "      <th>chr22</th>\n",
       "      <th>chrX</th>\n",
       "      <th>chrY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A06_0.sorted.markdup.bam</td>\n",
       "      <td>A06_0</td>\n",
       "      <td>0</td>\n",
       "      <td>788</td>\n",
       "      <td>1945</td>\n",
       "      <td>640</td>\n",
       "      <td>875</td>\n",
       "      <td>1105</td>\n",
       "      <td>477</td>\n",
       "      <td>998</td>\n",
       "      <td>...</td>\n",
       "      <td>275</td>\n",
       "      <td>494</td>\n",
       "      <td>317</td>\n",
       "      <td>229</td>\n",
       "      <td>341</td>\n",
       "      <td>305</td>\n",
       "      <td>196</td>\n",
       "      <td>274</td>\n",
       "      <td>707</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>B06_0.sorted.markdup.bam</td>\n",
       "      <td>B06_0</td>\n",
       "      <td>0</td>\n",
       "      <td>196758</td>\n",
       "      <td>350074</td>\n",
       "      <td>219828</td>\n",
       "      <td>144038</td>\n",
       "      <td>171211</td>\n",
       "      <td>119287</td>\n",
       "      <td>188950</td>\n",
       "      <td>...</td>\n",
       "      <td>71478</td>\n",
       "      <td>93397</td>\n",
       "      <td>82751</td>\n",
       "      <td>50048</td>\n",
       "      <td>81870</td>\n",
       "      <td>90936</td>\n",
       "      <td>55717</td>\n",
       "      <td>86933</td>\n",
       "      <td>135932</td>\n",
       "      <td>9549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D06_0.sorted.markdup.bam</td>\n",
       "      <td>D06_0</td>\n",
       "      <td>0</td>\n",
       "      <td>495209</td>\n",
       "      <td>713536</td>\n",
       "      <td>448474</td>\n",
       "      <td>334738</td>\n",
       "      <td>343263</td>\n",
       "      <td>240618</td>\n",
       "      <td>320277</td>\n",
       "      <td>...</td>\n",
       "      <td>150592</td>\n",
       "      <td>191267</td>\n",
       "      <td>173084</td>\n",
       "      <td>101996</td>\n",
       "      <td>173011</td>\n",
       "      <td>182343</td>\n",
       "      <td>114567</td>\n",
       "      <td>193443</td>\n",
       "      <td>272739</td>\n",
       "      <td>19480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>F11_1.sorted.markdup.bam</td>\n",
       "      <td>F11_1</td>\n",
       "      <td>1</td>\n",
       "      <td>219152</td>\n",
       "      <td>346247</td>\n",
       "      <td>220362</td>\n",
       "      <td>165834</td>\n",
       "      <td>171549</td>\n",
       "      <td>118818</td>\n",
       "      <td>190454</td>\n",
       "      <td>...</td>\n",
       "      <td>74473</td>\n",
       "      <td>95499</td>\n",
       "      <td>85291</td>\n",
       "      <td>50365</td>\n",
       "      <td>99103</td>\n",
       "      <td>91513</td>\n",
       "      <td>57702</td>\n",
       "      <td>96567</td>\n",
       "      <td>132343</td>\n",
       "      <td>9359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>D01_1.sorted.markdup.bam</td>\n",
       "      <td>D01_1</td>\n",
       "      <td>1</td>\n",
       "      <td>282699</td>\n",
       "      <td>511280</td>\n",
       "      <td>398890</td>\n",
       "      <td>272223</td>\n",
       "      <td>245213</td>\n",
       "      <td>171781</td>\n",
       "      <td>227863</td>\n",
       "      <td>...</td>\n",
       "      <td>107063</td>\n",
       "      <td>133457</td>\n",
       "      <td>120433</td>\n",
       "      <td>73151</td>\n",
       "      <td>122761</td>\n",
       "      <td>129033</td>\n",
       "      <td>80876</td>\n",
       "      <td>135745</td>\n",
       "      <td>191780</td>\n",
       "      <td>13955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>F10_1.sorted.markdup.bam</td>\n",
       "      <td>F10_1</td>\n",
       "      <td>1</td>\n",
       "      <td>218973</td>\n",
       "      <td>375966</td>\n",
       "      <td>245461</td>\n",
       "      <td>183436</td>\n",
       "      <td>190339</td>\n",
       "      <td>132794</td>\n",
       "      <td>176577</td>\n",
       "      <td>...</td>\n",
       "      <td>82958</td>\n",
       "      <td>104311</td>\n",
       "      <td>93206</td>\n",
       "      <td>56210</td>\n",
       "      <td>94104</td>\n",
       "      <td>101409</td>\n",
       "      <td>62239</td>\n",
       "      <td>104331</td>\n",
       "      <td>148289</td>\n",
       "      <td>10949</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        file sample  cell_count    chr1    chr2    chr3  \\\n",
       "4   A06_0.sorted.markdup.bam  A06_0           0     788    1945     640   \n",
       "6   B06_0.sorted.markdup.bam  B06_0           0  196758  350074  219828   \n",
       "0   D06_0.sorted.markdup.bam  D06_0           0  495209  713536  448474   \n",
       "59  F11_1.sorted.markdup.bam  F11_1           1  219152  346247  220362   \n",
       "56  D01_1.sorted.markdup.bam  D01_1           1  282699  511280  398890   \n",
       "11  F10_1.sorted.markdup.bam  F10_1           1  218973  375966  245461   \n",
       "\n",
       "      chr4    chr5    chr6    chr7  ...   chr15   chr16   chr17   chr18  \\\n",
       "4      875    1105     477     998  ...     275     494     317     229   \n",
       "6   144038  171211  119287  188950  ...   71478   93397   82751   50048   \n",
       "0   334738  343263  240618  320277  ...  150592  191267  173084  101996   \n",
       "59  165834  171549  118818  190454  ...   74473   95499   85291   50365   \n",
       "56  272223  245213  171781  227863  ...  107063  133457  120433   73151   \n",
       "11  183436  190339  132794  176577  ...   82958  104311   93206   56210   \n",
       "\n",
       "     chr19   chr20   chr21   chr22    chrX   chrY  \n",
       "4      341     305     196     274     707     79  \n",
       "6    81870   90936   55717   86933  135932   9549  \n",
       "0   173011  182343  114567  193443  272739  19480  \n",
       "59   99103   91513   57702   96567  132343   9359  \n",
       "56  122761  129033   80876  135745  191780  13955  \n",
       "11   94104  101409   62239  104331  148289  10949  \n",
       "\n",
       "[6 rows x 27 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# terminal command to get chromosome specific reads\n",
    "# samtools view -c -F 0x4 /file chr1 (chr2, chrX, chrY)\n",
    "x = 1\n",
    "while x <= 24:\n",
    "    if x == 23:\n",
    "        chr = \"chrX\"\n",
    "    elif x == 24:\n",
    "        chr = \"chrY\"\n",
    "    else:\n",
    "        chr = \"chr\" + str(x)\n",
    "\n",
    "    chr_list = []\n",
    "\n",
    "    for filename in test_df[\"file\"]:\n",
    "        chr_reads = subprocess.run([\"samtools\", \"view\", \"-c\", \"-F\", \"0x4\", directory + filename, chr], stdout=subprocess.PIPE, text=True)\n",
    "        chr_list.append(chr_reads.stdout[:-1])\n",
    "\n",
    "    test_df[chr] = chr_list\n",
    "    x += 1\n",
    "\n",
    "# optionally check the read counts mapping to each chromosome of each sample in the dataframe \n",
    "test_df #;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_count</th>\n",
       "      <th>chr1</th>\n",
       "      <th>chr2</th>\n",
       "      <th>chr3</th>\n",
       "      <th>chr4</th>\n",
       "      <th>chr5</th>\n",
       "      <th>chr6</th>\n",
       "      <th>chr7</th>\n",
       "      <th>...</th>\n",
       "      <th>chr16</th>\n",
       "      <th>chr17</th>\n",
       "      <th>chr18</th>\n",
       "      <th>chr19</th>\n",
       "      <th>chr20</th>\n",
       "      <th>chr21</th>\n",
       "      <th>chr22</th>\n",
       "      <th>chrX</th>\n",
       "      <th>chrY</th>\n",
       "      <th>total_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A06_0.sorted.markdup.bam</td>\n",
       "      <td>A06_0</td>\n",
       "      <td>0</td>\n",
       "      <td>788</td>\n",
       "      <td>1945</td>\n",
       "      <td>640</td>\n",
       "      <td>875</td>\n",
       "      <td>1105</td>\n",
       "      <td>477</td>\n",
       "      <td>998</td>\n",
       "      <td>...</td>\n",
       "      <td>494</td>\n",
       "      <td>317</td>\n",
       "      <td>229</td>\n",
       "      <td>341</td>\n",
       "      <td>305</td>\n",
       "      <td>196</td>\n",
       "      <td>274</td>\n",
       "      <td>707</td>\n",
       "      <td>79</td>\n",
       "      <td>13862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>B06_0.sorted.markdup.bam</td>\n",
       "      <td>B06_0</td>\n",
       "      <td>0</td>\n",
       "      <td>196758</td>\n",
       "      <td>350074</td>\n",
       "      <td>219828</td>\n",
       "      <td>144038</td>\n",
       "      <td>171211</td>\n",
       "      <td>119287</td>\n",
       "      <td>188950</td>\n",
       "      <td>...</td>\n",
       "      <td>93397</td>\n",
       "      <td>82751</td>\n",
       "      <td>50048</td>\n",
       "      <td>81870</td>\n",
       "      <td>90936</td>\n",
       "      <td>55717</td>\n",
       "      <td>86933</td>\n",
       "      <td>135932</td>\n",
       "      <td>9549</td>\n",
       "      <td>2985418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D06_0.sorted.markdup.bam</td>\n",
       "      <td>D06_0</td>\n",
       "      <td>0</td>\n",
       "      <td>495209</td>\n",
       "      <td>713536</td>\n",
       "      <td>448474</td>\n",
       "      <td>334738</td>\n",
       "      <td>343263</td>\n",
       "      <td>240618</td>\n",
       "      <td>320277</td>\n",
       "      <td>...</td>\n",
       "      <td>191267</td>\n",
       "      <td>173084</td>\n",
       "      <td>101996</td>\n",
       "      <td>173011</td>\n",
       "      <td>182343</td>\n",
       "      <td>114567</td>\n",
       "      <td>193443</td>\n",
       "      <td>272739</td>\n",
       "      <td>19480</td>\n",
       "      <td>6114118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>F11_1.sorted.markdup.bam</td>\n",
       "      <td>F11_1</td>\n",
       "      <td>1</td>\n",
       "      <td>219152</td>\n",
       "      <td>346247</td>\n",
       "      <td>220362</td>\n",
       "      <td>165834</td>\n",
       "      <td>171549</td>\n",
       "      <td>118818</td>\n",
       "      <td>190454</td>\n",
       "      <td>...</td>\n",
       "      <td>95499</td>\n",
       "      <td>85291</td>\n",
       "      <td>50365</td>\n",
       "      <td>99103</td>\n",
       "      <td>91513</td>\n",
       "      <td>57702</td>\n",
       "      <td>96567</td>\n",
       "      <td>132343</td>\n",
       "      <td>9359</td>\n",
       "      <td>3033992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>D01_1.sorted.markdup.bam</td>\n",
       "      <td>D01_1</td>\n",
       "      <td>1</td>\n",
       "      <td>282699</td>\n",
       "      <td>511280</td>\n",
       "      <td>398890</td>\n",
       "      <td>272223</td>\n",
       "      <td>245213</td>\n",
       "      <td>171781</td>\n",
       "      <td>227863</td>\n",
       "      <td>...</td>\n",
       "      <td>133457</td>\n",
       "      <td>120433</td>\n",
       "      <td>73151</td>\n",
       "      <td>122761</td>\n",
       "      <td>129033</td>\n",
       "      <td>80876</td>\n",
       "      <td>135745</td>\n",
       "      <td>191780</td>\n",
       "      <td>13955</td>\n",
       "      <td>4374589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>F10_1.sorted.markdup.bam</td>\n",
       "      <td>F10_1</td>\n",
       "      <td>1</td>\n",
       "      <td>218973</td>\n",
       "      <td>375966</td>\n",
       "      <td>245461</td>\n",
       "      <td>183436</td>\n",
       "      <td>190339</td>\n",
       "      <td>132794</td>\n",
       "      <td>176577</td>\n",
       "      <td>...</td>\n",
       "      <td>104311</td>\n",
       "      <td>93206</td>\n",
       "      <td>56210</td>\n",
       "      <td>94104</td>\n",
       "      <td>101409</td>\n",
       "      <td>62239</td>\n",
       "      <td>104331</td>\n",
       "      <td>148289</td>\n",
       "      <td>10949</td>\n",
       "      <td>3276243</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        file sample  cell_count    chr1    chr2    chr3  \\\n",
       "4   A06_0.sorted.markdup.bam  A06_0           0     788    1945     640   \n",
       "6   B06_0.sorted.markdup.bam  B06_0           0  196758  350074  219828   \n",
       "0   D06_0.sorted.markdup.bam  D06_0           0  495209  713536  448474   \n",
       "59  F11_1.sorted.markdup.bam  F11_1           1  219152  346247  220362   \n",
       "56  D01_1.sorted.markdup.bam  D01_1           1  282699  511280  398890   \n",
       "11  F10_1.sorted.markdup.bam  F10_1           1  218973  375966  245461   \n",
       "\n",
       "      chr4    chr5    chr6    chr7  ...   chr16   chr17   chr18   chr19  \\\n",
       "4      875    1105     477     998  ...     494     317     229     341   \n",
       "6   144038  171211  119287  188950  ...   93397   82751   50048   81870   \n",
       "0   334738  343263  240618  320277  ...  191267  173084  101996  173011   \n",
       "59  165834  171549  118818  190454  ...   95499   85291   50365   99103   \n",
       "56  272223  245213  171781  227863  ...  133457  120433   73151  122761   \n",
       "11  183436  190339  132794  176577  ...  104311   93206   56210   94104   \n",
       "\n",
       "     chr20   chr21   chr22    chrX   chrY total_reads  \n",
       "4      305     196     274     707     79       13862  \n",
       "6    90936   55717   86933  135932   9549     2985418  \n",
       "0   182343  114567  193443  272739  19480     6114118  \n",
       "59   91513   57702   96567  132343   9359     3033992  \n",
       "56  129033   80876  135745  191780  13955     4374589  \n",
       "11  101409   62239  104331  148289  10949     3276243  \n",
       "\n",
       "[6 rows x 28 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# terminal command to get total reads\n",
    "# samtools view -c -F 0x4 /file\n",
    "chr_list = []\n",
    "\n",
    "for filename in test_df[\"file\"]:\n",
    "        chr_reads = subprocess.run([\"samtools\", \"view\", \"-c\", \"-F\", \"0x4\", directory + filename], stdout=subprocess.PIPE, text=True)\n",
    "        chr_list.append(chr_reads.stdout[:-1])\n",
    "        \n",
    "test_df[\"total_reads\"] = chr_list\n",
    "\n",
    "# convert the numberic collumns in the dataframe to integers\n",
    "test_df[\"total_reads\"] = test_df[\"total_reads\"].astype(int)\n",
    "test_df[\"cell_count\"] = test_df[\"cell_count\"].astype(int)\n",
    "\n",
    "# optionally check the read counts mapping to each chromosome of each sample in the dataframe \n",
    "test_df #;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_count</th>\n",
       "      <th>chr1</th>\n",
       "      <th>chr2</th>\n",
       "      <th>chr3</th>\n",
       "      <th>chr4</th>\n",
       "      <th>chr5</th>\n",
       "      <th>chr6</th>\n",
       "      <th>chr7</th>\n",
       "      <th>...</th>\n",
       "      <th>chr15_percent</th>\n",
       "      <th>chr16_percent</th>\n",
       "      <th>chr17_percent</th>\n",
       "      <th>chr18_percent</th>\n",
       "      <th>chr19_percent</th>\n",
       "      <th>chr20_percent</th>\n",
       "      <th>chr21_percent</th>\n",
       "      <th>chr22_percent</th>\n",
       "      <th>chrX_percent</th>\n",
       "      <th>chrY_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A06_0.sorted.markdup.bam</td>\n",
       "      <td>A06_0</td>\n",
       "      <td>0</td>\n",
       "      <td>788</td>\n",
       "      <td>1945</td>\n",
       "      <td>640</td>\n",
       "      <td>875</td>\n",
       "      <td>1105</td>\n",
       "      <td>477</td>\n",
       "      <td>998</td>\n",
       "      <td>...</td>\n",
       "      <td>1.983841</td>\n",
       "      <td>3.563699</td>\n",
       "      <td>2.286827</td>\n",
       "      <td>1.651998</td>\n",
       "      <td>2.459962</td>\n",
       "      <td>2.200260</td>\n",
       "      <td>1.413937</td>\n",
       "      <td>1.976627</td>\n",
       "      <td>5.100274</td>\n",
       "      <td>0.569903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>B06_0.sorted.markdup.bam</td>\n",
       "      <td>B06_0</td>\n",
       "      <td>0</td>\n",
       "      <td>196758</td>\n",
       "      <td>350074</td>\n",
       "      <td>219828</td>\n",
       "      <td>144038</td>\n",
       "      <td>171211</td>\n",
       "      <td>119287</td>\n",
       "      <td>188950</td>\n",
       "      <td>...</td>\n",
       "      <td>2.394238</td>\n",
       "      <td>3.128440</td>\n",
       "      <td>2.771840</td>\n",
       "      <td>1.676415</td>\n",
       "      <td>2.742330</td>\n",
       "      <td>3.046006</td>\n",
       "      <td>1.866305</td>\n",
       "      <td>2.911921</td>\n",
       "      <td>4.553198</td>\n",
       "      <td>0.319855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D06_0.sorted.markdup.bam</td>\n",
       "      <td>D06_0</td>\n",
       "      <td>0</td>\n",
       "      <td>495209</td>\n",
       "      <td>713536</td>\n",
       "      <td>448474</td>\n",
       "      <td>334738</td>\n",
       "      <td>343263</td>\n",
       "      <td>240618</td>\n",
       "      <td>320277</td>\n",
       "      <td>...</td>\n",
       "      <td>2.463021</td>\n",
       "      <td>3.128284</td>\n",
       "      <td>2.830891</td>\n",
       "      <td>1.668205</td>\n",
       "      <td>2.829697</td>\n",
       "      <td>2.982327</td>\n",
       "      <td>1.873811</td>\n",
       "      <td>3.163874</td>\n",
       "      <td>4.460807</td>\n",
       "      <td>0.318607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>F11_1.sorted.markdup.bam</td>\n",
       "      <td>F11_1</td>\n",
       "      <td>1</td>\n",
       "      <td>219152</td>\n",
       "      <td>346247</td>\n",
       "      <td>220362</td>\n",
       "      <td>165834</td>\n",
       "      <td>171549</td>\n",
       "      <td>118818</td>\n",
       "      <td>190454</td>\n",
       "      <td>...</td>\n",
       "      <td>2.454621</td>\n",
       "      <td>3.147635</td>\n",
       "      <td>2.811181</td>\n",
       "      <td>1.660024</td>\n",
       "      <td>3.266423</td>\n",
       "      <td>3.016257</td>\n",
       "      <td>1.901851</td>\n",
       "      <td>3.182836</td>\n",
       "      <td>4.362009</td>\n",
       "      <td>0.308471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>D01_1.sorted.markdup.bam</td>\n",
       "      <td>D01_1</td>\n",
       "      <td>1</td>\n",
       "      <td>282699</td>\n",
       "      <td>511280</td>\n",
       "      <td>398890</td>\n",
       "      <td>272223</td>\n",
       "      <td>245213</td>\n",
       "      <td>171781</td>\n",
       "      <td>227863</td>\n",
       "      <td>...</td>\n",
       "      <td>2.447384</td>\n",
       "      <td>3.050732</td>\n",
       "      <td>2.753013</td>\n",
       "      <td>1.672180</td>\n",
       "      <td>2.806229</td>\n",
       "      <td>2.949603</td>\n",
       "      <td>1.848768</td>\n",
       "      <td>3.103034</td>\n",
       "      <td>4.383955</td>\n",
       "      <td>0.319001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>F10_1.sorted.markdup.bam</td>\n",
       "      <td>F10_1</td>\n",
       "      <td>1</td>\n",
       "      <td>218973</td>\n",
       "      <td>375966</td>\n",
       "      <td>245461</td>\n",
       "      <td>183436</td>\n",
       "      <td>190339</td>\n",
       "      <td>132794</td>\n",
       "      <td>176577</td>\n",
       "      <td>...</td>\n",
       "      <td>2.532108</td>\n",
       "      <td>3.183860</td>\n",
       "      <td>2.844905</td>\n",
       "      <td>1.715685</td>\n",
       "      <td>2.872314</td>\n",
       "      <td>3.095283</td>\n",
       "      <td>1.899706</td>\n",
       "      <td>3.184471</td>\n",
       "      <td>4.526191</td>\n",
       "      <td>0.334194</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        file sample  cell_count    chr1    chr2    chr3  \\\n",
       "4   A06_0.sorted.markdup.bam  A06_0           0     788    1945     640   \n",
       "6   B06_0.sorted.markdup.bam  B06_0           0  196758  350074  219828   \n",
       "0   D06_0.sorted.markdup.bam  D06_0           0  495209  713536  448474   \n",
       "59  F11_1.sorted.markdup.bam  F11_1           1  219152  346247  220362   \n",
       "56  D01_1.sorted.markdup.bam  D01_1           1  282699  511280  398890   \n",
       "11  F10_1.sorted.markdup.bam  F10_1           1  218973  375966  245461   \n",
       "\n",
       "      chr4    chr5    chr6    chr7  ...  chr15_percent  chr16_percent  \\\n",
       "4      875    1105     477     998  ...       1.983841       3.563699   \n",
       "6   144038  171211  119287  188950  ...       2.394238       3.128440   \n",
       "0   334738  343263  240618  320277  ...       2.463021       3.128284   \n",
       "59  165834  171549  118818  190454  ...       2.454621       3.147635   \n",
       "56  272223  245213  171781  227863  ...       2.447384       3.050732   \n",
       "11  183436  190339  132794  176577  ...       2.532108       3.183860   \n",
       "\n",
       "    chr17_percent  chr18_percent  chr19_percent  chr20_percent  chr21_percent  \\\n",
       "4        2.286827       1.651998       2.459962       2.200260       1.413937   \n",
       "6        2.771840       1.676415       2.742330       3.046006       1.866305   \n",
       "0        2.830891       1.668205       2.829697       2.982327       1.873811   \n",
       "59       2.811181       1.660024       3.266423       3.016257       1.901851   \n",
       "56       2.753013       1.672180       2.806229       2.949603       1.848768   \n",
       "11       2.844905       1.715685       2.872314       3.095283       1.899706   \n",
       "\n",
       "    chr22_percent  chrX_percent  chrY_percent  \n",
       "4        1.976627      5.100274      0.569903  \n",
       "6        2.911921      4.553198      0.319855  \n",
       "0        3.163874      4.460807      0.318607  \n",
       "59       3.182836      4.362009      0.308471  \n",
       "56       3.103034      4.383955      0.319001  \n",
       "11       3.184471      4.526191      0.334194  \n",
       "\n",
       "[6 rows x 52 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a collumn corresponding to the percentage of reads mapping to each chromosome\n",
    "total_read_list = [13862, 2985418, 6114118, 3033992, 4374589, 3276243] # FIXME: get rid of the manual imput of total reads for each sample\n",
    "\n",
    "x = 1\n",
    "while x <= 24:\n",
    "    if x == 23:\n",
    "        chr = \"chrX\"\n",
    "    elif x == 24:\n",
    "        chr = \"chrY\"\n",
    "    else:\n",
    "        chr = \"chr\" + str(x)\n",
    "\n",
    "    # convert the numeric collumns to integers\n",
    "    test_df[chr] = test_df[chr].astype(int)\n",
    "\n",
    "    chr_list = []\n",
    "    y = 0 # references back to the index of the total read list\n",
    "\n",
    "    for count in test_df[chr]:\n",
    "        chr_percent = (count / total_read_list[y]) * 100\n",
    "        chr_list.append(chr_percent)\n",
    "        y += 1\n",
    "    \n",
    "    test_df[chr + \"_percent\"] = chr_list\n",
    "    x += 1\n",
    "\n",
    "# optionally check the percentage of total reads mapping to each chromosome of each sample in the dataframe \n",
    "test_df #;   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A06_0 0 1 5.684605396046746\n",
      "B06_0 0 1 6.590634879269838\n",
      "D06_0 0 1 8.099434783561586\n",
      "F11_1 1 1 7.223222737568194\n",
      "D01_1 1 1 6.462298515357672\n",
      "F10_1 1 1 6.683661743039207\n",
      "A06_0 0 2 14.031164334150917\n",
      "B06_0 0 2 11.726130143249621\n",
      "D06_0 0 2 11.670301423688583\n",
      "F11_1 1 2 11.412258173390041\n",
      "D01_1 1 2 11.687497956950928\n",
      "F10_1 1 2 11.47552242004027\n",
      "A06_0 0 3 4.616938392728322\n",
      "B06_0 0 3 7.36339098913452\n",
      "D06_0 0 3 7.335056340096806\n",
      "F11_1 1 3 7.263104187486322\n",
      "D01_1 1 3 9.118342317415419\n",
      "F10_1 1 3 7.4921487813938095\n",
      "A06_0 0 4 6.312220458808253\n",
      "B06_0 0 4 4.824718012687001\n",
      "D06_0 0 4 5.474837090157567\n",
      "F11_1 1 4 5.465868070845276\n",
      "D01_1 1 4 6.222824589921476\n",
      "F10_1 1 4 5.598974190864353\n",
      "A06_0 0 5 7.971432693694994\n",
      "B06_0 0 5 5.73490881343919\n",
      "D06_0 0 5 5.6142684848411495\n",
      "F11_1 1 5 5.6542337619875065\n",
      "D01_1 1 5 5.6053951582651536\n",
      "F10_1 1 5 5.809672847832105\n",
      "A06_0 0 6 3.4410618958303276\n",
      "B06_0 0 6 3.9956548798191744\n",
      "D06_0 0 6 3.9354490704955314\n",
      "F11_1 1 6 3.9162265424562754\n",
      "D01_1 1 6 3.9267917511793677\n",
      "F10_1 1 6 4.053240251104695\n",
      "A06_0 0 7 7.199538306160727\n",
      "B06_0 0 7 6.329096963976235\n",
      "D06_0 0 7 5.238318920243279\n",
      "F11_1 1 7 6.277340217113295\n",
      "D01_1 1 7 5.208786471140489\n",
      "F10_1 1 7 5.389618535621442\n",
      "A06_0 0 8 7.26446400230847\n",
      "B06_0 0 8 6.201275667259995\n",
      "D06_0 0 8 5.211593888112725\n",
      "F11_1 1 8 6.063430622097883\n",
      "D01_1 1 8 5.109485713972216\n",
      "F10_1 1 8 5.334067100639361\n",
      "A06_0 0 9 2.611455778386957\n",
      "B06_0 0 9 3.406122693706543\n",
      "D06_0 0 9 3.3428370208098697\n",
      "F11_1 1 9 2.6029073247391556\n",
      "D01_1 1 9 3.330278570169678\n",
      "F10_1 1 9 3.39513888316587\n",
      "A06_0 0 10 3.6574808829894674\n",
      "B06_0 0 10 4.31259542214859\n",
      "D06_0 0 10 4.301716126512442\n",
      "F11_1 1 10 4.2528127958148865\n",
      "D01_1 1 10 4.221676596361395\n",
      "F10_1 1 10 4.408586298391175\n",
      "A06_0 0 11 3.1452892800461694\n",
      "B06_0 0 11 4.516185003239078\n",
      "D06_0 0 11 4.474774611808277\n",
      "F11_1 1 11 4.500110745183243\n",
      "D01_1 1 11 4.4448518477964445\n",
      "F10_1 1 11 4.574752239073841\n",
      "A06_0 0 12 5.107488096955707\n",
      "B06_0 0 12 4.6733824208201336\n",
      "D06_0 0 12 4.6114909787478755\n",
      "F11_1 1 12 4.6192936566741105\n",
      "D01_1 1 12 4.612981928130848\n",
      "F10_1 1 12 4.697392714765052\n",
      "A06_0 0 13 3.5420574231712596\n",
      "B06_0 0 13 1.990709508685216\n",
      "D06_0 0 13 1.9623599021150722\n",
      "F11_1 1 13 1.7529710032195205\n",
      "D01_1 1 13 1.9525262830405326\n",
      "F10_1 1 13 2.0098020812253545\n",
      "A06_0 0 14 2.0848362429663827\n",
      "B06_0 0 14 2.6140058109115705\n",
      "D06_0 0 14 2.618464347596824\n",
      "F11_1 1 14 2.6482930739435044\n",
      "D01_1 1 14 2.5961524614083746\n",
      "F10_1 1 14 2.678037007633439\n",
      "A06_0 0 15 1.9838407156254507\n",
      "B06_0 0 15 2.3942375908499245\n",
      "D06_0 0 15 2.4630208314592554\n",
      "F11_1 1 15 2.4546208427708445\n",
      "D01_1 1 15 2.447384199978558\n",
      "F10_1 1 15 2.532107661122817\n",
      "A06_0 0 16 3.563699321887174\n",
      "B06_0 0 16 3.128439635588718\n",
      "D06_0 0 16 3.128284406679753\n",
      "F11_1 1 16 3.147635194819235\n",
      "D01_1 1 16 3.0507323087951805\n",
      "F10_1 1 16 3.1838602936351177\n",
      "A06_0 0 17 2.286827297648247\n",
      "B06_0 0 17 2.7718396552844524\n",
      "D06_0 0 17 2.830890735180446\n",
      "F11_1 1 17 2.811180780964485\n",
      "D01_1 1 17 2.7530129116129536\n",
      "F10_1 1 17 2.8449049719449993\n",
      "A06_0 0 18 1.6519982686481027\n",
      "B06_0 0 18 1.6764151619639192\n",
      "D06_0 0 18 1.6682046372019643\n",
      "F11_1 1 18 1.6600241529971076\n",
      "D01_1 1 18 1.672179946504689\n",
      "F10_1 1 18 1.715684703485059\n",
      "A06_0 0 19 2.459962487375559\n",
      "B06_0 0 19 2.742329549831883\n",
      "D06_0 0 19 2.8296967771966455\n",
      "F11_1 1 19 3.266422587798518\n",
      "D01_1 1 19 2.806229339487664\n",
      "F10_1 1 19 2.872314416238356\n",
      "A06_0 0 20 2.200259702784591\n",
      "B06_0 0 20 3.0460056179737642\n",
      "D06_0 0 20 2.982327132057314\n",
      "F11_1 1 20 3.016257129221171\n",
      "D01_1 1 20 2.949602808400972\n",
      "F10_1 1 20 3.0952832253285245\n",
      "A06_0 0 21 1.4139373827730486\n",
      "B06_0 0 21 1.8663048189566753\n",
      "D06_0 0 21 1.8738107442479848\n",
      "F11_1 1 21 1.9018507629552088\n",
      "D01_1 1 21 1.8487679642590424\n",
      "F10_1 1 21 1.8997064625548226\n",
      "A06_0 0 22 1.9766267493868128\n",
      "B06_0 0 22 2.911920541780079\n",
      "D06_0 0 22 3.1638741679503077\n",
      "F11_1 1 22 3.1828363423502766\n",
      "D01_1 1 22 3.1030343650569234\n",
      "F10_1 1 22 3.1844707489645914\n",
      "A06_0 0 X 5.100274130717068\n",
      "B06_0 0 X 4.553198245605808\n",
      "D06_0 0 X 4.46080693895669\n",
      "F11_1 1 X 4.36200886488824\n",
      "D01_1 1 X 4.383954698372807\n",
      "F10_1 1 X 4.52619051761423\n",
      "A06_0 0 Y 0.5699033328524022\n",
      "B06_0 0 Y 0.3198547071130408\n",
      "D06_0 0 Y 0.3186068701977947\n",
      "F11_1 1 Y 0.3084714791601296\n",
      "D01_1 1 Y 0.3190013964740459\n",
      "F10_1 1 Y 0.3341937701202261\n"
     ]
    }
   ],
   "source": [
    "# use this code snippet to rearrange the data\n",
    "# opened the full output with a text editor and saved file\n",
    "x = 1\n",
    "while x <= 24:\n",
    "    if x == 23:\n",
    "        chr = \"chrX_percent\"\n",
    "    elif x == 24:\n",
    "        chr = \"chrY_percent\"\n",
    "    else:\n",
    "        chr = \"chr\" + str(x) + \"_percent\"\n",
    "\n",
    "    for index in test_df.index:\n",
    "        if len(chr) == 12:\n",
    "            print((test_df[\"sample\"][index]), test_df[\"cell_count\"][index], chr[3:4], test_df[chr][index])\n",
    "        elif len(chr) == 13:\n",
    "            print((test_df[\"sample\"][index]), test_df[\"cell_count\"][index], chr[3:5], test_df[chr][index])\n",
    "    \n",
    "    x+=1\n",
    "\n",
    "# saved to sc_chr_mapping_df.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the data into permanent storage\n",
    "test_df.to_csv(\"/Users/David.Sokol/Documents/experiments/20220718_scwgbs_contamination_analysis/data/edited_test_df.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.8",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "4b1d709454758155bdbdfdae4f21284fdbfe3f291544079d2527636be25f1e41"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
