{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "\n",
    "# generate an empty dataframe to be filled with each sample sequenced in the experiment\n",
    "sample_df = pd.DataFrame()\n",
    "\n",
    "# assign the file path to the experiments sequencing data\n",
    "directory = \"/put/path/to/directory/containing/bam/files/here\"\n",
    "\n",
    "# iterate over each file in the directory, remove duplicate names, and populate lists with a minimal sample name as well as the cell counts\n",
    "file_list = []\n",
    "sample_list = []\n",
    "cell_count_list = []\n",
    "\n",
    "for filename in os.listdir(directory):\n",
    "    if \"sorted.markdup.bam.bai\" in filename:\n",
    "        cell_count_list.append(1)\n",
    "        file_list.append(filename[:-3])\n",
    "        sample_list.append(filename[:5]) # CHANGEME: QC this step to be sure you are getting what you want\n",
    "\n",
    "# fill the sample data frame in with the list generated from the experiment directory\n",
    "sample_df[\"file\"] = sc_file_list\n",
    "sample_df[\"sample\"] = sc_sample_list\n",
    "sample_df[\"cell_count\"] = sc_cell_count_list\n",
    "\n",
    "# randomly take a subset of the samples sent to sequencing in order to save CPU and time\n",
    "test_df = sample_df.sample(frac=0.5)\n",
    "\n",
    "# read the data into permanent storage and optionally check the samples in the test dataframe\n",
    "test_df.to_csv(\"/put/path/to/directory/containing/experiment/data/here\")\n",
    "test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# terminal command to get chromosome specific reads\n",
    "# samtools view -c -F 0x4 /file chr1 (chr2, chrX, chrY)\n",
    "x = 1\n",
    "while x <= 24:\n",
    "    if x == 23:\n",
    "        chr = \"chrX\"\n",
    "    elif x == 24:\n",
    "        chr = \"chrY\"\n",
    "    else:\n",
    "        chr = \"chr\" + str(x)\n",
    "\n",
    "    chr_list = []\n",
    "\n",
    "    for filename in test_df[\"file\"]:\n",
    "        chr_reads = subprocess.run([\"samtools\", \"view\", \"-c\", \"-F\", \"0x4\", directory + filename, chr], stdout=subprocess.PIPE, text=True)\n",
    "        chr_list.append(chr_reads.stdout[:-1])\n",
    "\n",
    "    test_df[chr] = chr_list\n",
    "    x += 1\n",
    "\n",
    "# optionally check the read counts mapping to each chromosome of each sample in the dataframe \n",
    "test_df #;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# terminal command to get total reads\n",
    "# samtools view -c -F 0x4 /file\n",
    "chr_list = []\n",
    "\n",
    "for filename in test_df[\"file\"]:\n",
    "        chr_reads = subprocess.run([\"samtools\", \"view\", \"-c\", \"-F\", \"0x4\", directory + filename], stdout=subprocess.PIPE, text=True)\n",
    "        chr_list.append(chr_reads.stdout[:-1])\n",
    "        \n",
    "test_df[\"total_reads\"] = chr_list\n",
    "\n",
    "# convert the numberic collumns in the dataframe to integers\n",
    "test_df[\"total_reads\"] = test_df[\"total_reads\"].astype(int)\n",
    "test_df[\"cell_count\"] = test_df[\"cell_count\"].astype(int)\n",
    "\n",
    "# optionally check the read counts mapping to each chromosome of each sample in the dataframe \n",
    "test_df #;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a collumn corresponding to the percentage of reads mapping to each chromosome\n",
    "total_read_list = [put, total, reads, here] # FIXME: get rid of the manual imput of total reads for each sample\n",
    "\n",
    "x = 1\n",
    "while x <= 24:\n",
    "    if x == 23:\n",
    "        chr = \"chrX\"\n",
    "    elif x == 24:\n",
    "        chr = \"chrY\"\n",
    "    else:\n",
    "        chr = \"chr\" + str(x)\n",
    "\n",
    "    # convert the numeric collumns to integers\n",
    "    test_df[chr] = test_df[chr].astype(int)\n",
    "\n",
    "    chr_list = []\n",
    "    y = 0 # references back to the index of the total read list\n",
    "\n",
    "    for count in test_df[chr]:\n",
    "        chr_percent = (count / total_read_list[y]) * 100\n",
    "        chr_list.append(chr_percent)\n",
    "        y += 1\n",
    "    \n",
    "    test_df[chr + \"_percent\"] = chr_list\n",
    "    x += 1\n",
    "\n",
    "# optionally check the percentage of total reads mapping to each chromosome of each sample in the dataframe \n",
    "test_df #;   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use this code snippet to rearrange the data\n",
    "# opened the full output with a text editor and saved file\n",
    "x = 1\n",
    "while x <= 24:\n",
    "    if x == 23:\n",
    "        chr = \"chrX_percent\"\n",
    "    elif x == 24:\n",
    "        chr = \"chrY_percent\"\n",
    "    else:\n",
    "        chr = \"chr\" + str(x) + \"_percent\"\n",
    "\n",
    "    for index in test_df.index:\n",
    "        if len(chr) == 12:\n",
    "            print((test_df[\"sample\"][index]), test_df[\"cell_count\"][index], chr[3:4], test_df[chr][index])\n",
    "        elif len(chr) == 13:\n",
    "            print((test_df[\"sample\"][index]), test_df[\"cell_count\"][index], chr[3:5], test_df[chr][index])\n",
    "    \n",
    "    x+=1"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
